### ISSUE TITLE
Implement scikit-learn compatibility

### ISSUE THREAD

**Issue by jake-westfall** [bambinos/bambi#105]
It would be nice to be able to easily get posterior predictive distributions for arbitrary predictor values. The most obvious way to implement this would be to add a `.predict()` method to either the `Model` or `ModelResults` class. This would also bring the interface one little step closer to being `scikit-learn`-compatible, which is a separate goal that I think it would be nice to eventually achieve.

**Comment by aflaxman** [comment]
Thanks for all of your work on this package, I think it will be very useful for me!  Here is a small example of how I would like the `.predict()` method to work:

```
import numpy as np, pandas as pd, pymc3 as pm, bambi

df = pd.DataFrame({'p':np.random.normal(size=400)})

# blank out some values, for which I would like to impute
df.loc[np.random.choice(df.index, size=200, replace=False), 'p'] = np.nan

model = bambi.Model(df, dropna=True)
results = model.fit('p ~ 1')

p_pred = model.predict(df)  # something like pm.sample_ppc(model.backend.trace, model=model.backend.model, samples=1000), but including the rows with p == nan
p_pred.shape = (400, 1000)
```

**Comment by tyarkoni** [comment]
I like both the idea of implementing a `.predict` method and of doing PPCs, but I'm not sure the former should just wrap the latter. Naively, I think many people used to the sklearn API from other contexts would be surprised in the above example to get back (a) more than one column and (b) a non-deterministic result. Also, I think we'd need to allow the user to control the number of samples, and it feels a bit contrary to the spirit of the sklearn API to allow arguments to `predict` (e.g., for specification of what predict should do and/or how many posterior samples to draw).

I'm not suggesting that `predict` *should* just return a single column in the above case (it's not obvious what that would be anyway--the MAP?), but if we're going to break the standard sklearn patterns, we should probably have a good justification for that. The alternative is just to implement a `sample_ppc` method on the base `Model` class (which might require some work in the Stan backend if PyStan doesn't already support PPC sampling) and defer making a decision on a `predict` API for now.

**Comment by jake-westfall** [comment]
Maybe the wrapping should be the reverse: add a `.sample_ppc()` method (although I'm not sure if I love that name) that is the main prediction method and does all the real work, and then afterward add a `.predict()` method that just wraps `.sample_ppc()` and calls it with some sensible defaults that make it compatible with `scikit-learn`. And if supplying arguments to `.predict()` would make `scikit-learn` angry (e.g., not supported in Pipelines), then maybe have some other `Model` method that alters the behavior of `.predict()`.

**Comment by jake-westfall** [comment]
Just realized that having `predict()` wrap `sample_ppc` is exactly what you said you weren't sure about :p

**Comment by jake-westfall** [comment]
Still riffing here, but another idea is to have no `.predict()` method by default, but add a separate method that instantiates a `.predict()` method with specified behavior. So if you wanted to put a bambi model is a `sciki-learn` Pipeline you'd have to first specify how `.predict()` would work, but there'd at least be some nice functionality for doing so painlessly.

**Comment by tyarkoni** [comment]
I like that idea, and it's similar to a pattern I've used in other packages--wrapping the library's native class(es) in a class that implements sklearn's `BaseEstimator` interface. If we did that, we could have a `BambiSKLearnEstimator` class that takes a fitted bambi `Model` in init, along with arguments specifiying how to behave when `.fit` is called. I think that would be consistent with the sklearn API and wouldn't be too cumbersome. Then @aflaxman's example might look like:

```python
import numpy as np, pandas as pd, pymc3 as pm, bambi

df = pd.DataFrame({'p':np.random.normal(size=400)})

# blank out some values, for which I would like to impute
df.loc[np.random.choice(df.index, size=200, replace=False), 'p'] = np.nan

model = bambi.Model(df, dropna=True)
results = model.fit('p ~ 1')
est = bambi.BambiSKLearnEstimator(model, return='ppc', samples=1000)
p_pred = est.predict(df)
assert p_pred.shape == (400, 1000)
```

This still isn't completely right, because I don't see any obvious way we could implement `fit` in an sklearn-like way--our models can't be constructed without already having the data. But it's probably the closest thing.

**Comment by jake-westfall** [comment]
Yes, I was just about to write the same idea. So then we could defer all `sklearn` compatibility issues until we have a fully working `BambiSKLearnEstimator` class.

As for the `fit` behavior, we can pre-compile the model without any data, right? Stan definitely does this, and it's a basic part of the `RStan` interface, so I'm guessing PyStan supports it too (but can't remember). I'm guessing PyMC3 also supports compiling the model without any data, although I can't immediately see or remember how that works there.

**Comment by aflaxman** [comment]
The recently revamped `GaussianProcessRegression` module in `sklearn` could be helpful inspiration.  In addition to `.predict`, this class has a `.sample_y` method, which takes `n_samples` and `random_state` are arguments. https://github.com/scikit-learn/scikit-learn/blob/01e0639b15d515b446cb4589f01097b16fb4994a/sklearn/gaussian_process/gpr.py#L349

So fun to get to see your careful thinking about implementation!

**Comment by tyarkoni** [comment]
Re: `fit`, PyMC3 and Stan can both precompile, but unless I'm mistaken, the shapes of the parameters have to be known when compiling (at least in PyMC3--not sure about Stan off-hand). So this wouldn't work in a case where, e.g., the number of levels in a categorical variable isn't known in advance--which seems like a pretty big constraint.

**Comment by jake-westfall** [comment]
In Stan you pass in the shapes of the parameters as part of the data -- so you do need to know the numbers of dimensions for the parameters, but not the _sizes_ of each dimension. For PyMC3 it looks like we'd have to think hard about how/if we could make that work.

**Comment by jake-westfall** [comment]
Maybe we should break off this discussion into a new thread about a `BambiSKLearnEstimator` class.

**Comment by tyarkoni** [comment]
I guess we could always refactor the `Model` code so that it accepts a model specification in `fit` but does nothing else until it's passed some data. That would give us full sklearn compatibility, but (a) it would be a breaking change, as it would require no longer passing the data at initialization, and (b) it would be much slower, since the compilation would have to be done inside the `fit` call, which defeats the general idea of separating init from fit.

Given these complications, I'm not really inclined to make big changes just to get sklearn compatibility. Let's keep thinking about this, but maybe open a separate issue for PPC sampling, which should happen regardless.

**Comment by tyarkoni** [comment]
I think we can use this thread since the discussion is already here... will rename it.

**Comment by jake-westfall** [comment]
Yes, PPC functionality should happen regardless.

**Comment by tomicapretto** [comment]
In-sample and out-of-sample predictions have been implemented in #372
